File: eval-reviews/2026-09-12-chained/workflow-recovery/typescript/fable/r1/starter/leased.ts
/**
 * Checkpoint two: leased workers with fenced recovery. Selected when the
 * request carries a `workers` field; the checkpoint-one simulator in
 * workflow.ts is otherwise used unchanged.
 */
import {
  DomainError,
  MockService,
  TIME_LIMIT,
  fields,
  identifier,
  integer,
  parseStep,
  require,
  validateGraph,
  type ActionKey,
  type Effect,
  type ExecuteOutcome,
  type LookupOutcome,
  type ServiceCall,
  type StepDefinition,
} from "./workflow.ts";

export const MAX_COMMANDS = 2_000;
export const MAX_WORKERS = 100;
export const MAX_STEPS = 200;

export type LeasedCommand =
  | { op: "start" | "cancel"; run: string }
  | { op: "advance"; by: number }
  | { op: "observe" }
  | { op: "crash" | "restart" | "claim"; worker: string }
  | { op: "renew" | "call"; worker: string; ticket: number }
  | { op: "deliver"; ticket: number };

export interface LeasedWorkflow {
  readonly steps: readonly StepDefinition[];
  readonly workers: readonly string[];
  readonly commands: readonly LeasedCommand[];
  readonly maxAttempts: number;
  readonly retryDelay: number;
  readonly leaseDuration: number;
}

function parseCommand(raw: unknown): LeasedCommand {
  require(raw !== null && typeof raw === "object" && !Array.isArray(raw));
  const obj = raw as Record<string, unknown>;
  switch (obj.op) {
    case "start":
    case "cancel":
      fields(raw, ["op", "run"]);
      return { op: obj.op, run: identifier(obj.run) };
    case "advance":
      fields(raw, ["op", "by"]);
      return { op: obj.op, by: integer(obj.by, 0, TIME_LIMIT) };
    case "observe":
      fields(raw, ["op"]);
      return { op: obj.op };
    case "crash":
    case "restart":
    case "claim":
      fields(raw, ["op", "worker"]);
      return { op: obj.op, worker: identifier(obj.worker) };
    case "renew":
    case "call":
      fields(raw, ["op", "worker", "ticket"]);
      return {
        op: obj.op,
        worker: identifier(obj.worker),
        ticket: integer(obj.ticket, 1, TIME_LIMIT),
      };
    case "deliver":
      fields(raw, ["op", "ticket"]);
      return { op: obj.op, ticket: integer(obj.ticket, 1, TIME_LIMIT) };
    default:
      throw new DomainError("INVALID_INPUT");
  }
}

export function parseLeasedWorkflow(raw: unknown): LeasedWorkflow {
  const obj = fields(
    raw,
    ["steps", "commands", "workers"],
    ["max_attempts", "retry_delay", "lease_duration"],
  );
  require(
    Array.isArray(obj.steps) &&
      obj.steps.length > 0 &&
      obj.steps.length <= MAX_STEPS &&
      Array.isArray(obj.commands) &&
      obj.commands.length <= MAX_COMMANDS &&
      Array.isArray(obj.workers) &&
      obj.workers.length > 0 &&
      obj.workers.length <= MAX_WORKERS,
  );
  const workers = obj.workers.map(identifier);
  require(new Set(workers).size === workers.length);
  const optional = (key: string, fallback: number, low: number, high: number) =>
    integer(Object.hasOwn(obj, key) ? obj[key] : fallback, low, high);
  const workflow: LeasedWorkflow = {
    steps: obj.steps.map(parseStep),
    workers,
    commands: obj.commands.map(parseCommand),
    maxAttempts: optional("max_attempts", 3, 1, 10),
    retryDelay: optional("retry_delay", 2, 1, 1_000_000),
    leaseDuration: optional("lease_duration", 5, 1, 1_000_000),
  };
  validateGraph(workflow.steps); // Shape/scalar validation precedes graph checks.
  return workflow;
}

export type LeasedStepStatus =
  | "pending"
  | "running"
  | "succeeded"
  | "failed"
  | "blocked"
  | "cancelled";
export type LeasedRunStatus =
  | "active"
  | "succeeded"
  | "failed"
  | "cancelling"
  | "cancelled"
  | "failing";
export interface Lease {
  worker: string;
  ticket: number;
  expires: number;
}
export interface LeasedStepState {
  id: string;
  status: LeasedStepStatus;
  attempts: number;
  ready_at: number;
  lease: Lease | null;
}
export interface LeasedRunState {
  id: string;
  status: LeasedRunStatus;
  cancel_requested: boolean;
  steps: LeasedStepState[];
}
export interface WorkerState {
  id: string;
  up: boolean;
}
export type Receipt =
  | { kind: "execute"; outcome: ExecuteOutcome }
  | { kind: "lookup"; outcome: LookupOutcome };
/** Durable metadata for one lease acquisition, retained after replacement. */
interface Ticket {
  readonly ticket: number;
  readonly worker: string;
  readonly runIndex: number;
  readonly stepIndex: number;
  readonly kind: "execute" | "lookup";
  readonly attempt: number;
  response: Receipt | null; // Delayed transport message; survives crashes.
  delivered: boolean;
}
export interface LeasedSnapshot {
  now: number;
  workers: WorkerState[];
  runs: LeasedRunState[];
}
export type CommandValue =
  | null
  | LeasedSnapshot
  | { ticket: number | null }
  | { renewed: boolean }
  | { committed: boolean }
  | Receipt
  | { outcome: "stale" };
export interface LeasedResult {
  results: CommandValue[];
  final: LeasedSnapshot & { calls: ServiceCall[]; effects: Effect[] };
}

const NO_CANCEL: ReadonlySet<LeasedRunStatus> = new Set([
  "succeeded",
  "failed",
  "cancelled",
  "failing",
]);

export class LeasedSimulator {
  readonly workflow: LeasedWorkflow;
  now = 0;
  readonly workers: WorkerState[];
  readonly runs: LeasedRunState[] = [];
  readonly service: MockService;
  private readonly tickets = new Map<number, Ticket>();
  private nextTicket = 1;
  constructor(workflow: LeasedWorkflow) {
    this.workflow = workflow;
    this.workers = workflow.workers.map((id) => ({ id, up: true }));
    this.service = new MockService(workflow.steps);
  }
  private live(lease: Lease | null): lease is Lease {
    return lease !== null && this.now < lease.expires;
  }
  private worker(id: string): WorkerState {
    const worker = this.workers.find((candidate) => candidate.id === id);
    require(worker !== undefined, "UNKNOWN_WORKER");
    return worker;
  }
  private upWorker(id: string): WorkerState {
    const worker = this.worker(id);
    require(worker.up, "WORKER_DOWN");
    return worker;
  }
  private ownedTicket(worker: string, id: number): Ticket {
    const ticket = this.tickets.get(id);
    require(ticket !== undefined, "UNKNOWN_TICKET");
    require(ticket.worker === worker, "WRONG_WORKER");
    return ticket;
  }
  private stepOf(ticket: Ticket): LeasedStepState {
    return this.runs[ticket.runIndex]!.steps[ticket.stepIndex]!;
  }
  /** True when the ticket is the step's current lease and that lease is live. */
  private current(ticket: Ticket): boolean {
    const lease = this.stepOf(ticket).lease;
    return this.live(lease) && lease.ticket === ticket.ticket;
  }
  private deadline(base: number, delta: number): number {
    require(base + delta <= TIME_LIMIT, "TIME_OVERFLOW");
    return base + delta;
  }
  /** Revoke every running lease of a run so earlier responses become stale. */
  private expireRunning(run: LeasedRunState): void {
    for (const step of run.steps)
      if (step.status === "running" && step.lease) step.lease.expires = this.now;
  }
  private selectWork(): { runIndex: number; stepIndex: number } | undefined {
    for (const [runIndex, run] of this.runs.entries())
      for (const [stepIndex, step] of run.steps.entries())
        if (step.status === "running" && !this.live(step.lease))
          return { runIndex, stepIndex };
    for (const [runIndex, run] of this.runs.entries()) {
      if (run.status !== "active") continue;
      const succeeded = new Set(
        run.steps
          .filter((step) => step.status === "succeeded")
          .map((step) => step.id),
      );
      for (const [stepIndex, step] of run.steps.entries()) {
        const definition = this.workflow.steps[stepIndex]!;
        if (
          step.status === "pending" &&
          step.ready_at <= this.now &&
          definition.needs.every((dep) => succeeded.has(dep))
        )
          return { runIndex, stepIndex };
      }
    }
    return undefined;
  }
  claim(workerId: string): { ticket: number | null } {
    const worker = this.upWorker(workerId);
    require(
      !this.runs.some((run) =>
        run.steps.some(
          (step) => this.live(step.lease) && step.lease.worker === worker.id,
        ),
      ),
      "WORKER_BUSY",
    );
    const work = this.selectWork();
    if (!work) return { ticket: null };
    const run = this.runs[work.runIndex]!;
    const step = run.steps[work.stepIndex]!;
    const expires = this.deadline(this.now, this.workflow.leaseDuration);
    if (step.status === "pending") {
      step.status = "running";
      step.attempts += 1;
    }
    const ticket: Ticket = {
      ticket: this.nextTicket++,
      worker: worker.id,
      runIndex: work.runIndex,
      stepIndex: work.stepIndex,
      kind: run.status === "active" ? "execute" : "lookup",
      attempt: step.attempts,
      response: null,
      delivered: false,
    };
    this.tickets.set(ticket.ticket, ticket);
    step.lease = { worker: worker.id, ticket: ticket.ticket, expires };
    return { ticket: ticket.ticket };
  }
  renew(workerId: string, id: number): { renewed: boolean } {
    const worker = this.upWorker(workerId);
    const ticket = this.ownedTicket(worker.id, id);
    if (!this.current(ticket)) return { renewed: false };
    this.stepOf(ticket).lease!.expires = this.deadline(
      this.now,
      this.workflow.leaseDuration,
    );
    return { renewed: true };
  }
  call(workerId: string, id: number): Receipt | { outcome: "stale" } {
    const worker = this.upWorker(workerId);
    const ticket = this.ownedTicket(worker.id, id);
    if (!this.current(ticket)) return { outcome: "stale" };
    if (ticket.response) return { ...ticket.response }; // Saved receipt, no audit.
    const run = this.runs[ticket.runIndex]!;
    const step = this.stepOf(ticket);
    const key: ActionKey = [run.id, step.id];
    const origin = { worker: ticket.worker, ticket: ticket.ticket };
    ticket.response =
      ticket.kind === "execute"
        ? {
            kind: "execute",
            outcome: this.service.execute(
              key,
              this.workflow.steps[ticket.stepIndex]!.amount,
              ticket.attempt,
              origin,
            ),
          }
        : {
            kind: "lookup",
            outcome: this.service.lookup(key, ticket.attempt, origin),
          };
    return { ...ticket.response };
  }
  deliver(id: number): { committed: boolean } {
    const ticket = this.tickets.get(id);
    require(ticket !== undefined, "UNKNOWN_TICKET");
    if (
      !ticket.response ||
      ticket.delivered ||
      !this.current(ticket) ||
      !this.worker(ticket.worker).up
    )
      return { committed: false };
    const run = this.runs[ticket.runIndex]!;
    const step = this.stepOf(ticket);
    const response = ticket.response;
    if (response.kind === "execute" && response.outcome === "transient") {
      if (step.attempts < this.workflow.maxAttempts) {
        const readyAt = this.deadline(this.now, this.workflow.retryDelay);
        ticket.delivered = true;
        step.lease = null;
        step.status = "pending";
        step.ready_at = readyAt;
        return { committed: true };
      }
      ticket.delivered = true;
      step.lease = null;
      step.status = "failed";
      for (const other of run.steps)
        if (other.status === "pending") other.status = "blocked";
      if (run.steps.some((other) => other.status === "running")) {
        run.status = "failing";
        this.expireRunning(run);
      } else run.status = "failed";
      return { committed: true };
    }
    ticket.delivered = true;
    step.lease = null;
    if (response.kind === "execute") {
      step.status = "succeeded";
      if (
        run.status === "active" &&
        run.steps.every((other) => other.status === "succeeded")
      )
        run.status = "succeeded";
      return { committed: true };
    }
    const draining = run.status === "failing";
    step.status =
      response.outcome === "found"
        ? "succeeded"
        : draining
          ? "blocked"
          : "cancelled";
    if (!run.steps.some((other) => other.status === "running"))
      run.status = draining ? "failed" : "cancelled";
    return { committed: true };
  }
  cancel(run: LeasedRunState): void {
    if (NO_CANCEL.has(run.status) || run.status === "cancelling") return;
    run.cancel_requested = true;
    for (const step of run.steps)
      if (step.status === "pending") step.status = "cancelled";
    if (run.steps.some((step) => step.status === "running")) {
      run.status = "cancelling";
      this.expireRunning(run); // First cancellation only; lookups keep leases.
    } else run.status = "cancelled";
  }
  apply(command: LeasedCommand): CommandValue {
    switch (command.op) {
      case "start":
        require(
          this.runs.every((run) => run.id !== command.run),
          "DUPLICATE_RUN",
        );
        this.runs.push({
          id: command.run,
          status: "active",
          cancel_requested: false,
          steps: this.workflow.steps.map((step) => ({
            id: step.id,
            status: "pending",
            attempts: 0,
            ready_at: this.now,
            lease: null,
          })),
        });
        return null;
      case "advance":
        this.now = this.deadline(this.now, command.by);
        return null;
      case "observe":
        return this.snapshot();
      case "crash":
        this.upWorker(command.worker).up = false;
        return null;
      case "restart": {
        const worker = this.worker(command.worker);
        require(!worker.up, "WORKER_UP");
        worker.up = true; // Passive: recovery happens through later claims.
        return null;
      }
      case "claim":
        return this.claim(command.worker);
      case "renew":
        return this.renew(command.worker, command.ticket);
      case "call":
        return this.call(command.worker, command.ticket);
      case "deliver":
        return this.deliver(command.ticket);
      case "cancel": {
        const run = this.runs.find((candidate) => candidate.id === command.run);
        require(run !== undefined, "UNKNOWN_RUN");
        this.cancel(run);
        return null;
      }
    }
  }
  snapshot(): LeasedSnapshot {
    return {
      now: this.now,
      workers: this.workers.map((worker) => ({ ...worker })),
      runs: this.runs.map((run) => ({
        ...run,
        steps: run.steps.map((step) => ({
          ...step,
          lease: step.lease ? { ...step.lease } : null,
        })),
      })),
    };
  }
  run(): LeasedResult {
    const results = this.workflow.commands.map((command) => this.apply(command));
    return {
      results,
      final: {
        ...this.snapshot(),
        calls: this.service.calls,
        effects: this.service.effects,
      },
    };
  }
}

File: eval-reviews/2026-09-12-chained/workflow-recovery/python/fable/r1/starter/leased.py
"""Checkpoint two: leased workers, delayed calls/commits, fencing, and failure draining.

This module implements the `workers` input mode. The checkpoint-one simulator in
`workflow.py` is untouched; the shared validation helpers, step parsing, graph checks,
and the idempotent mock service are reused from it.
"""
from dataclasses import dataclass, field
from typing import Any, Literal

from workflow import (TIME_LIMIT, DomainError, MockService, StepDefinition, identifier,
                      integer, object_fields, parse_step, require, validate_graph)

COMMAND_LIMIT = 2_000


@dataclass(frozen=True)
class LeasedCommand:
    op: str
    run: str | None = None
    by: int = 0
    worker: str | None = None
    ticket: int | None = None


@dataclass(frozen=True)
class LeasedWorkflow:
    steps: tuple[StepDefinition, ...]
    commands: tuple[LeasedCommand, ...]
    workers: tuple[str, ...]
    max_attempts: int
    retry_delay: int
    lease_duration: int


def parse_leased_command(raw: Any) -> LeasedCommand:
    require(type(raw) is dict and type(raw.get("op")) is str)
    op = raw["op"]
    if op in ("start", "cancel"):
        object_fields(raw, {"op", "run"})
        return LeasedCommand(op, run=identifier(raw["run"]))
    if op == "advance":
        object_fields(raw, {"op", "by"})
        return LeasedCommand(op, by=integer(raw["by"], 0, TIME_LIMIT))
    if op == "observe":
        object_fields(raw, {"op"})
        return LeasedCommand(op)
    if op in ("crash", "restart", "claim"):
        object_fields(raw, {"op", "worker"})
        return LeasedCommand(op, worker=identifier(raw["worker"]))
    if op in ("renew", "call"):
        object_fields(raw, {"op", "worker", "ticket"})
        return LeasedCommand(op, worker=identifier(raw["worker"]),
                             ticket=integer(raw["ticket"], 1, TIME_LIMIT))
    require(op == "deliver")
    object_fields(raw, {"op", "ticket"})
    return LeasedCommand(op, ticket=integer(raw["ticket"], 1, TIME_LIMIT))


def parse_leased_workflow(raw: Any) -> LeasedWorkflow:
    obj = object_fields(raw, {"steps", "commands", "workers"},
                        {"max_attempts", "retry_delay", "lease_duration"})
    require(type(obj["steps"]) is list and bool(obj["steps"]))
    require(type(obj["commands"]) is list and len(obj["commands"]) <= COMMAND_LIMIT)
    require(type(obj["workers"]) is list and bool(obj["workers"]))
    steps = tuple(parse_step(step) for step in obj["steps"])
    workers = tuple(identifier(worker) for worker in obj["workers"])
    require(len(workers) == len(set(workers)))
    commands = tuple(parse_leased_command(command) for command in obj["commands"])
    workflow = LeasedWorkflow(steps, commands, workers,
                              integer(obj.get("max_attempts", 3), 1, 10),
                              integer(obj.get("retry_delay", 2), 1, 1_000_000),
                              integer(obj.get("lease_duration", 5), 1, 1_000_000))
    validate_graph(steps)  # Field/scalar validation completes before graph checks.
    return workflow


@dataclass
class Lease:
    worker: str
    ticket: int
    expires: int


@dataclass
class LeasedStep:
    id: str
    status: Literal["pending", "running", "succeeded", "failed", "blocked", "cancelled"] = "pending"
    attempts: int = 0
    ready_at: int = 0
    lease: Lease | None = None


@dataclass
class LeasedRun:
    id: str
    steps: list[LeasedStep]
    status: Literal["active", "succeeded", "failed", "cancelling", "cancelled", "failing"] = "active"
    cancel_requested: bool = False

    @property
    def terminal(self) -> bool:
        return self.status in ("succeeded", "failed", "cancelled")

    def has_running(self) -> bool:
        return any(step.status == "running" for step in self.steps)


@dataclass
class Ticket:
    """Durable acquisition metadata plus the delayed transport message (saved response)."""
    worker: str
    run: LeasedRun
    step: LeasedStep
    definition: StepDefinition
    kind: Literal["execute", "lookup"]
    attempt: int
    response: str | None = None
    delivered: bool = False


@dataclass
class AuditedService(MockService):
    """The checkpoint-one mock plus worker/ticket attribution for each audited call."""
    origins: list[tuple[str, int]] = field(default_factory=list)

    def execute_as(self, worker: str, ticket: int, key: tuple[str, str], amount: int,
                   attempt: int) -> str:
        self.origins.append((worker, ticket))
        return self.execute(key, amount, attempt)

    def lookup_as(self, worker: str, ticket: int, key: tuple[str, str], attempt: int) -> str:
        self.origins.append((worker, ticket))
        return self.lookup(key, attempt)


class LeasedSimulator:
    def __init__(self, workflow: LeasedWorkflow):
        self.workflow = workflow
        self.now = 0
        self.workers: dict[str, bool] = {worker: True for worker in workflow.workers}
        self.runs: list[LeasedRun] = []
        self.tickets: dict[int, Ticket] = {}
        self.next_ticket = 1
        self.service = AuditedService({step.id: step.failures for step in workflow.steps})
        self.results: list[Any] = []

    # ----- lookups and validation -------------------------------------------------
    def find_run(self, run_id: str) -> LeasedRun:
        for run in self.runs:
            if run.id == run_id:
                return run
        raise DomainError("UNKNOWN_RUN")

    def require_worker(self, worker: str, up: bool = True) -> None:
        require(worker in self.workers, "UNKNOWN_WORKER")
        if up:
            require(self.workers[worker], "WORKER_DOWN")

    def require_ticket(self, worker: str, ticket: int) -> Ticket:
        require(ticket in self.tickets, "UNKNOWN_TICKET")
        record = self.tickets[ticket]
        require(record.worker == worker, "WRONG_WORKER")
        return record

    def is_live(self, lease: Lease | None) -> bool:
        return lease is not None and self.now < lease.expires

    def is_current_live(self, ticket: int, record: Ticket) -> bool:
        lease = record.step.lease
        return lease is not None and lease.ticket == ticket and self.is_live(lease)

    def lease_deadline(self) -> int:
        deadline = self.now + self.workflow.lease_duration
        require(deadline <= TIME_LIMIT, "TIME_OVERFLOW")
        return deadline

    # ----- scheduling --------------------------------------------------------------
    def select_work(self) -> tuple[LeasedRun, LeasedStep, StepDefinition] | None:
        for run in self.runs:
            for state, definition in zip(run.steps, self.workflow.steps):
                if state.status == "running" and not self.is_live(state.lease):
                    return run, state, definition
        for run in self.runs:
            if run.status != "active":
                continue
            succeeded = {step.id for step in run.steps if step.status == "succeeded"}
            for state, definition in zip(run.steps, self.workflow.steps):
                if (state.status == "pending" and state.ready_at <= self.now
                        and all(dep in succeeded for dep in definition.needs)):
                    return run, state, definition
        return None

    def claim(self, worker: str) -> dict:
        self.require_worker(worker)
        for run in self.runs:
            for step in run.steps:
                lease = step.lease
                require(not (lease is not None and lease.worker == worker
                             and self.is_live(lease)), "WORKER_BUSY")
        work = self.select_work()
        if work is None:
            return {"ticket": None}
        run, state, definition = work
        expires = self.lease_deadline()
        ticket = self.next_ticket
        self.next_ticket += 1
        if state.status == "pending":
            state.status = "running"
            state.attempts += 1
        kind = "execute" if run.status == "active" else "lookup"
        state.lease = Lease(worker, ticket, expires)
        self.tickets[ticket] = Ticket(worker, run, state, definition, kind, state.attempts)
        return {"ticket": ticket}

    def renew(self, worker: str, ticket: int) -> dict:
        self.require_worker(worker)
        record = self.require_ticket(worker, ticket)
        if not self.is_current_live(ticket, record):
            return {"renewed": False}
        record.step.lease.expires = self.lease_deadline()
        return {"renewed": True}

    def call(self, worker: str, ticket: int) -> dict:
        self.require_worker(worker)
        record = self.require_ticket(worker, ticket)
        if not self.is_current_live(ticket, record):
            return {"outcome": "stale"}
        if record.response is None:
            key = (record.run.id, record.step.id)
            if record.kind == "execute":
                record.response = self.service.execute_as(worker, ticket, key,
                                                          record.definition.amount, record.attempt)
            else:
                record.response = self.service.lookup_as(worker, ticket, key, record.attempt)
        return {"kind": record.kind, "outcome": record.response}

    # ----- commit -----------------------------------------------------------------
    def deliver(self, ticket: int) -> dict:
        require(ticket in self.tickets, "UNKNOWN_TICKET")
        record = self.tickets[ticket]
        if (record.response is None or record.delivered or not self.workers[record.worker]
                or not self.is_current_live(ticket, record)):
            return {"committed": False}
        record.delivered = True
        run, step = record.run, record.step
        step.lease = None
        if record.kind == "lookup":
            if record.response == "found":
                step.status = "succeeded"
            else:
                step.status = "cancelled" if run.status == "cancelling" else "blocked"
            if not run.has_running():
                run.status = "cancelled" if run.status == "cancelling" else "failed"
        elif record.response in ("applied", "replayed"):
            step.status = "succeeded"
            if all(s.status == "succeeded" for s in run.steps):
                run.status = "succeeded"
        elif step.attempts < self.workflow.max_attempts:
            deadline = self.now + self.workflow.retry_delay
            require(deadline <= TIME_LIMIT, "TIME_OVERFLOW")
            step.status = "pending"
            step.ready_at = deadline
        else:
            step.status = "failed"
            for other in run.steps:
                if other.status == "pending":
                    other.status = "blocked"
            if run.has_running():
                run.status = "failing"
                self.expire_running(run)
            else:
                run.status = "failed"
        return {"committed": True}

    def expire_running(self, run: LeasedRun) -> None:
        for step in run.steps:
            if step.status == "running" and step.lease is not None:
                step.lease.expires = self.now

    def cancel(self, run: LeasedRun) -> None:
        if run.status != "active":
            return  # Terminal, failing, and repeated cancellations are no-ops.
        run.cancel_requested = True
        for step in run.steps:
            if step.status == "pending":
                step.status = "cancelled"
        if run.has_running():
            run.status = "cancelling"
            self.expire_running(run)
        else:
            run.status = "cancelled"

    # ----- command dispatch -------------------------------------------------------
    def apply(self, command: LeasedCommand) -> Any:
        op = command.op
        if op == "start":
            require(all(run.id != command.run for run in self.runs), "DUPLICATE_RUN")
            self.runs.append(LeasedRun(command.run, [LeasedStep(step.id, ready_at=self.now)
                                                     for step in self.workflow.steps]))
            return None
        if op == "advance":
            require(self.now + command.by <= TIME_LIMIT, "TIME_OVERFLOW")
            self.now += command.by
            return None
        if op == "observe":
            return self.snapshot()
        if op == "crash":
            self.require_worker(command.worker)
            self.workers[command.worker] = False
            return None
        if op == "restart":
            self.require_worker(command.worker, up=False)
            require(not self.workers[command.worker], "WORKER_UP")
            self.workers[command.worker] = True
            return None
        if op == "claim":
            return self.claim(command.worker)
        if op == "renew":
            return self.renew(command.worker, command.ticket)
        if op == "call":
            return self.call(command.worker, command.ticket)
        if op == "deliver":
            return self.deliver(command.ticket)
        if op == "cancel":
            self.cancel(self.find_run(command.run))
            return None
        raise DomainError("INVALID_INPUT")

    def snapshot(self) -> dict:
        # Fresh containers at every level keep earlier results immutable.
        return {
            "now": self.now,
            "workers": [{"id": worker, "up": up} for worker, up in self.workers.items()],
            "runs": [{"id": run.id, "status": run.status, "cancel_requested": run.cancel_requested,
                      "steps": [{"id": step.id, "status": step.status, "attempts": step.attempts,
                                 "ready_at": step.ready_at,
                                 "lease": None if step.lease is None else {
                                     "worker": step.lease.worker, "ticket": step.lease.ticket,
                                     "expires": step.lease.expires}}
                                for step in run.steps]} for run in self.runs],
        }

    def run(self) -> dict:
        for command in self.workflow.commands:
            self.results.append(self.apply(command))
        final = self.snapshot()
        final["calls"] = [{"worker": worker, "ticket": ticket, "kind": call.kind,
                           "key": list(call.key), "attempt": call.attempt, "outcome": call.outcome}
                          for (worker, ticket), call in zip(self.service.origins, self.service.calls)]
        final["effects"] = [{"key": list(effect.key), "amount": effect.amount}
                            for effect in self.service.effects]
        return {"results": self.results, "final": final}
